pkg/issuer/vault/issue.go

```diff
@@ -130,7 +132,80 @@ func (v *Vault) configureCertPool(cfg *vault.Config) error {
 	return nil
 }
 
+var (
+	clientRepo   map[*v1alpha1.VaultIssuer]*vault.Client
+	clientRepoMu sync.Mutex
+)
+
+func renewToken(vaultClient *vault.Client) {
+	resp, err := vaultClient.Auth().Token().LookupSelf()
+	if err != nil {
+		glog.Warningf("vault: lookup-self failed, token renewal is disabled: %s", err)
+		return
+	}
+
+	b, _ := json.Marshal(resp.Data)
+	var data struct {
+		TTL         int       `json:"ttl"`
+		CreationTTL int       `json:"creation_ttl"`
+		Renewable   bool      `json:"renewable"`
+		ExpireTime  time.Time `json:"expire_time"`
+	}
+	if err := json.Unmarshal(b, &data); err != nil {
+		glog.Warningf("vault: lookup-self failed, token renewal is disabled: %s", err)
+		return
+	}
+
+	switch {
+	case data.Renewable:
+		// renewable token, proceed further
+	case data.ExpireTime.IsZero():
+		// nothing to do. token doesn't expire
+		return
+	default:
+		ttl := time.Until(data.ExpireTime) / time.Second * time.Second // truncate to secs
+		glog.Warningf("vault: Token is not renewable and it will expire %s from now at %s",
+			ttl, data.ExpireTime.Format(time.RFC3339))
+		return
+	}
+
+	ttl := time.Duration(data.TTL) * time.Second
+	timer := time.NewTimer(ttl / 2)
+
+	for range timer.C {
+		resp, err := vaultClient.Auth().Token().RenewSelf(data.CreationTTL)
+		if err != nil {
+			vaultClient.ClearToken()
+			glog.Warningf("vault: Failed to renew token: %s. Removed it from client", err)
+			return
+		}
+
+		if !resp.Auth.Renewable || resp.Auth.LeaseDuration == 0 {
+			glog.Infof("vault: token is no longer renewable so removed it from client")
+			vaultClient.ClearToken()
+			return
+		}
+		ttl = time.Duration(resp.Auth.LeaseDuration) * time.Second
```

Comment:
what's this line doing? Some comments here would be great 😄

Reply:
It reads the new ttl after token has been renewed. yes, i will add some comments.